resdk/resolwe.py

```diff
@@ -92,115 +84,88 @@ def print_process_inputs(self, process_name):
-            raise ValueError('Data must be dict, str or unicode.')
-
-        resource = resource.lower()
-        if resource not in ('data', 'collection', 'process', 'trigger', 'template'):
-            raise ValueError('Resource must be data, collection, process, trigger or template.')
-
-        url = urljoin(self.url, '/api/{}'.format(resource))
-        return requests.post(url,
-                             data=data,
-                             auth=self.auth,
-                             headers={
-                                 'cache-control': 'no-cache',
-                                 'content-type': 'application/json',
-                                 'accept': 'application/json, text/plain, */*',
-                                 'referer': self.url,
-                             })
-
-    def upload(self, process_name, name='', descriptor=None, descriptor_schema=None,
-               collections=[], **fields):
-
-        """Upload files and data objects.
-
-        :param collections: Integer id of Resolwe collection
-        :type collections: List of int
-        :param process_name: Processor object name
-        :type process_name: string
-        :param fields: Processor field-value pairs
-        :type fields: args
+    def run(self, slug=None, input={}, descriptor=None,  # pylint: disable=redefined-builtin
+            descriptor_schema=None, collections=[], data_name=''):
+        """Run process and return the corresponding Data object.
+
+        1. Upload files referenced in inputs
+        2. Create Data object with given inputs
+        3. Command is run to process inputs into outputs
+        4. Return Data object
+
+        The processing runs asynchronously, so the returned Data object
+        does not have an OK status nor outputs yet.
+
+        :param slug: Process slug (unique identifier)
+        :type slug: str
+        :param input: Input values
+        :type input: dict
+        :param descriptor: Descriptor values
+        :type descriptor: dict
+        :param descriptor_schema: Descriptor fields
+        :type descriptor_schema: list of dicts
+        :param collections: Default collections of Data object
+        :type collections: list of ints
+        :param data_name: Default name of Data object
+        :type data_name: string
         :rtype: HTTP Response object
 
         """
-        # This is temporary solution: map process name to it's ID:
-        proc_name_to_slug = dict([(x['name'], x['slug']) for x in self.processes()])
+        if ((descriptor and not descriptor_schema) or
+                (not descriptor and descriptor_schema)):
+            raise ValueError("Set both or neither descriptor and descriptor_schema")
 
-        process = self.processes(process_name=process_name)
+        process = self.api.process.get(slug=slug, ordering='-version', limit=1)
 
         if len(process) == 1:
             process = process[0]
-        elif len(process) > 1:
```

Comment:
This block of code handles the situation when there are two versions of one process (two equal slugs for two different versions of same process). It just ensures that the latest version of processor is taken. Multiple versions of processes exist becouse of backward compatibility (on genesis), at least i was told so by @dblenkus.
Anyway, I think this part should be kept.